src/serac/physics/tests/CMakeLists.txt

```diff
@@ -0,0 +1,57 @@
+# Copyright (c) 2019-2021, Lawrence Livermore National Security, LLC and
+# other Serac Project Developers. See the top-level LICENSE file for
+# details.
+#
+# SPDX-License-Identifier: (BSD-3-Clause) 
+
+if (ENABLE_GTEST)
+
+    set(test_dependencies serac_physics test_utils)
+    blt_list_append( TO test_dependencies ELEMENTS caliper IF ${SERAC_USE_CALIPER} )
+
+    blt_add_library(
+        NAME        test_utils
+        SOURCES     test_utilities.cpp
+        HEADERS     test_utilities.hpp
+        DEPENDS_ON  serac_physics gtest
+        )
+
+    set(solver_tests
+        serac_solid.cpp
+        serac_solid_reuse_mesh.cpp
+        serac_thermal_solver.cpp
+        serac_thermal_structural_solver.cpp
+        )
+
+    foreach(filename ${solver_tests})
+        get_filename_component(test_name ${filename} NAME_WE)
+
+        blt_add_executable(NAME        ${test_name}
+                           SOURCES     ${filename}
+                           OUTPUT_DIR  ${TEST_OUTPUT_DIRECTORY}
+                           DEPENDS_ON  ${test_dependencies}
+                           FOLDER      serac/tests )
+        blt_add_test(NAME          ${test_name}
+                     COMMAND       ${test_name}
+                     NUM_MPI_TASKS 2 )
+    endforeach()
+
+    set(solver_utility_tests
+        serac_dtor.cpp
+        serac_boundary_cond.cpp
+        serac_newmark_test.cpp)
+
+    foreach(filename ${solver_utility_tests})
+        get_filename_component(test_name ${filename} NAME_WE)
+
+        blt_add_executable(NAME        ${test_name}
+                           SOURCES     ${filename}
+                           OUTPUT_DIR  ${TEST_OUTPUT_DIRECTORY}
+                           DEPENDS_ON  ${test_dependencies}
+                           FOLDER      serac/tests )
+        blt_add_test(NAME          ${test_name}
+                     COMMAND       ${test_name}
+                     NUM_MPI_TASKS 2 )
+    endforeach()
```

Comment:
this is another option to avoid us having to explicitly write out the names of each file in every CMakeLists.txt
```cmake
    file(GLOB tests_in_this_directory "serac_*.cpp")
 
    foreach(filename ${tests_in_this_directory})
        get_filename_component(test_name ${filename} NAME_WE)
 
         blt_add_executable(NAME        ${test_name}
                             SOURCES     ${filename}
                             OUTPUT_DIR  ${TEST_OUTPUT_DIRECTORY}
                             DEPENDS_ON  ${test_dependencies}
                             FOLDER      serac/tests )
         blt_add_test(NAME          ${test_name}
                       COMMAND       ${test_name}
                       NUM_MPI_TASKS 2 )
     endforeach()
```
it might also be nice to have an additional function `add_mpi_test()` to replace this all with
```cmake
file(GLOB tests_in_this_directory "serac_*.cpp")
foreach(filename ${tests_in_this_directory})
  add_mpi_test(${filename} NUM_MPI_TASKS 2)
endforeach()
```

but I know developers people feel strongly about never using `file(GLOB ... )`

Reply:
> 
> 
> I think this is an improvement, but it's making me realize how much boilerplate we have in all of our CMakeLists.txt. For example, this snippet is used everywhere, verbatim:
> 
> ```cmake
> foreach(filename ${whatever_tests})
>      get_filename_component(test_name ${filename} NAME_WE)
> 
>      blt_add_executable(NAME        ${test_name}
>                         SOURCES     ${filename}
>                         OUTPUT_DIR  ${TEST_OUTPUT_DIRECTORY}
>                         DEPENDS_ON  ${test_dependencies}
>                         FOLDER      serac/tests )
>      blt_add_test(NAME          ${test_name}
>                   COMMAND       ${test_name}
>                   NUM_MPI_TASKS 1 )
>  endforeach()
> ```
> 
> it might be worth making a function to avoid copy-and-pasting that same body in every single file, e.g.
> 
> ```cmake
> foreach(filename ${whatever_tests})
>   add_serial_test(filename FOLDER serac/tests)
> endforeach()
> ```
> 
> Additionally, it seems like there's another pattern where some_directory/CMakeLists.txt has
> 
> ```cmake
> ...
> if (enable_testing)
>   add_subdirectory(tests)
> endif()
> ...
> ```
> 
> and then some_directory/tests/CMakeLists.txt has another conditional guard
> 
> ```cmake
> if (ENABLE_GTEST)
> ...
> # entire contents of file
> ...
> endif()
> ```
> 
> Is there a case where we ever enable testing, but don't have gtest?

Thoughts on this @white238 ?

Reply:
Globing is not recommended for source files by anyone in the CMake community (even the creators) it causes CMake to lose track of source dependencies.  They semi-recently added a `CONFIGURE_DEPENDS` option to `file(GLOB ..)` but they still recommend against it due to the rebuild time cost of rerunning that command everytime.

see the note here: https://cmake.org/cmake/help/latest/command/file.html#glob